### ISSUE TITLE
IDLE cond context: fix code update and font update timers

### ISSUE THREAD

**Issue by terryjreedy** [python/cpython#75674]
BPO | [31493](https://bugs.python.org/issue31493)
--- | :---
Nosy | @terryjreedy, @csabella
PRs | <li>python/cpython#3622</li><li>python/cpython#3623</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/terryjreedy'
closed_at = <Date 2018-05-23.05:25:54.854>
created_at = <Date 2017-09-16.20:47:15.593>
labels = ['expert-IDLE', 'type-bug', '3.7']
title = 'IDLE cond context: fix code update and font update timers'
updated_at = <Date 2018-05-23.05:25:54.853>
user = 'https://github.com/terryjreedy'
```

bugs.python.org fields:
```python
activity = <Date 2018-05-23.05:25:54.853>
actor = 'terry.reedy'
assignee = 'terry.reedy'
closed = True
closed_date = <Date 2018-05-23.05:25:54.854>
closer = 'terry.reedy'
components = ['IDLE']
creation = <Date 2017-09-16.20:47:15.593>
creator = 'terry.reedy'
dependencies = []
files = []
hgrepos = []
issue_num = 31493
keywords = ['patch']
message_count = 6.0
messages = ['302352', '302359', '302360', '302628', '302706', '317360']
nosy_count = 2.0
nosy_names = ['terry.reedy', 'cheryl.sabella']
pr_nums = ['3622', '3623']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue31493'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by terryjreedy** [comment]
1. Currently, each instance of codecontext.CodeContext creates 2 root.after loops.  There should be at most two loops for the class, and preferably fewer by using the new reload().

2. The loops are never explicitly shut down.  Since conversion to a feature,  the following appears after running test_idle.

.invalid command name "109891384timer_event"
    while executing
"109891384timer_event"
    ("after" script)
invalid command name "109891512font_timer_event"
    while executing
"109891512font_timer_event"
    ("after" script)

This is because of
  File "F:\\dev\\3x\\lib\\idlelib\\idle_test\\test_outwin.py", line 20, in setUpClass
    w = cls.window = outwin.OutputWindow(None, None, None, root)
  File "F:\\dev\\3x\\lib\\idlelib\\outwin.py", line 78, in \_\_init__
    EditorWindow.\_\_init__(self, \*args)
  File "F:\\dev\\3x\\lib\\idlelib\\editor.py", line 319, in \_\_init__
  text.bind("\<\<toggle-code-context\>\>",  # added to traceback
    self.CodeContext(self).toggle_code_context_event)

The statement, which creates an instance, was added as part of the conversion.  However, the outwin tests were merged a week before the conversion, so this message might have started the.  In any case, editor.EditorWindow.load_extension does the same thing in doing the binding.  It is not clear why there is no message from other editor-creating tests. 

I want to fix at least this before next Monday's cutoff for new releases.  I will first try is after_cancel in a \_\_del__ method.

**Comment by terryjreedy** [comment]
New changeset a6bb313c70f8619e6dc4af5cef7d73fa3bbd59ca by Terry Jan Reedy in branch 'master':
bpo-31493: Fix code context update and font update timers. (bpo-3622)
https://github.com/python/cpython/commit/a6bb313c70f8619e6dc4af5cef7d73fa3bbd59ca

**Comment by terryjreedy** [comment]
New changeset b417332bf44af93c8acda4b7197181d4b62c16a4 by Terry Jan Reedy (Miss Islington (bot)) in branch '3.6':
[3.6] bpo-31493: Fix code context update and font update timers. (GH-3622) (bpo-3623)
https://github.com/python/cpython/commit/b417332bf44af93c8acda4b7197181d4b62c16a4

**Comment by csabella** [comment]
Sorry, I didn't see this until now.

I added a line to editor.py in bpo-31529 for this error message.  In _close() in editor.py, self.text was being set to None which wasn't calling the __del__ function of multicall.  I added a line to unbind "<<toggle-code-context>>", which made the error go away.  I don't know if that's the proper fix or not, but it seemed the error was occurring because the `after` wasn't being removed at the right time.

**Comment by terryjreedy** [comment]
The underlying issue is trying to coordinate two somewhat independent object systems (Python and Tcl).  I don't know why the after loops were (apparently) stopped before bpo-27099 and not after.  The only difference I know of is the dropping of the reference to CodeContext instances in .unload_extensions, called from .\_close.  CodeContext did not have a .close method.

I tried the first thing I thought might work and it did, so I stopped.  I don't know why unbinding the toggle function was enough.  I expect that we will eventually review .\_close.  It should help when the menu and editor are separated a bit.
